### PR TITLE
Add support for running tests on non-empty backend systems

### PR DESCRIPTION
**Pull request by jorjich**

### CODE REVIEW

File: scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/E2EGroupComplianceTest.java [hunk 1]
```diff
@@ -95,29 +99,31 @@ void testGetGroupsEmptyList() {
...
-    return () -> assertAll("Verify memberRefs are same ",
-        expectedMembers.stream()
-            .map(MemberRef::getValue)
-            .map(expMemberVal ->
-              () -> assertTrue(actualMembers.stream().map(MemberRef::getValue).anyMatch(expMemberVal::equals), "Verify member existence"))
-        );
+  @TestFactory
+  @DisableOnEmptyGroupsEndpoint
+  public Collection<DynamicTest> testGetMultipleExistingGroups() {
+    return Arrays.asList(getMultipleGroupsDynamicTest("Test get multiple groups", () -> {
```
Comment (karaimin) [inline scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/E2EGroupComplianceTest.java]: `Collections.singletonList()`

File: scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/E2EGroupComplianceTest.java [hunk 2]
```diff
@@ -314,10 +319,8 @@ public void testUpdateGroupRemovingOnlyOneMember() {
     logger.info("Creating Group -{}- with members", groupDisplayName);
     Group createdTestGroupWithMembers = groupFailSafeClient.create(buildGroup(groupDisplayName, firstUserMember.getId()));
 
-    Group groupForUpdate = new Group.Builder(createdTestGroupWithMembers)
-        .removeMembers()
-        .addMember(buildGroupMemberResourceWithId(firstUserMember.getId()))
-        .build();
+    Group groupForUpdate = new Group.Builder(createdTestGroupWithMembers).removeMembers()
```
Comment (karaimin) [inline scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/E2EGroupComplianceTest.java]: surround with formatter and restore its position

File: scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/E2EGroupComplianceTest.java [hunk 3]
```diff
@@ -685,11 +689,9 @@ public void testGroupMemberDisplayNameUpdateWithPUTRequest() throws InterruptedE
     List<String> groupMembersDisplayNames = getGroupMembersDisplayNames(createdTestGroupWithMembers, updateMemberDisplayNameTestUserNewName,
         updateMemberDisplayNameTestGroupNewName);
 
-    assertAll("Verify member.display is updated in parent group",
-        () -> assertEquals(2, groupMembersDisplayNames.size(), "Verify members size"),
+    assertAll("Verify member.display is updated in parent group", () -> assertEquals(2, groupMembersDisplayNames.size(), "Verify members size"),
```
Comment (karaimin) [inline scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/E2EGroupComplianceTest.java]: surround with formatter and restore its position

File: scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/E2EGroupComplianceTest.java [hunk 4]
```diff
@@ -727,11 +729,9 @@ public void testGroupMemberDisplayNameUpdateWithPATCHRequest() throws Interrupte
 
     List<String> groupMembersDisplayNames = getGroupMembersDisplayNames(createdTestGroupWithMembers, newUserDisplayName, newGroupDisplayName);
 
-    assertAll("Verify member.display is updated in parent group",
-        () -> assertEquals(2, groupMembersDisplayNames.size(), "Verify members size"),
+    assertAll("Verify member.display is updated in parent group", () -> assertEquals(2, groupMembersDisplayNames.size(), "Verify members size"),
```
Comment (karaimin) [inline scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/E2EGroupComplianceTest.java]: surround with formatter and restore its position

File: scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/E2EGroupComplianceTest.java [hunk 5]
```diff
@@ -805,6 +806,7 @@ public void testGetGroupsWithStarIndexEqualTotalResults() {
 
   @Test
   @DisplayName("Test Get groups with index paging and startIndex out of range (more than total results)")
+  @EnableOnEmptyGroupsEndpoint
```
Comment (karaimin) [inline scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/E2EGroupComplianceTest.java]: Similar body. Extract it.

File: scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/E2EGroupComplianceTest.java [hunk 6]
```diff
@@ -780,6 +780,7 @@ public void testGetGroupsTotalCountWithStartId() {
 
   @Test
   @DisplayName("Test Get groups with index paging and startIndex=totalResults")
+  @EnableOnEmptyGroupsEndpoint
```
Comment (karaimin) [inline scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/E2EGroupComplianceTest.java]: Write the same test for non-empty backend

File: scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/E2EGroupComplianceTest.java [hunk 7]
```diff
@@ -970,6 +996,7 @@ public void testGetGroupsPagingStartIdWithUpperCase() {
 
   @Test
   @DisplayName("Test Get groups all pages with index paging")
+  @EnableOnEmptyGroupsEndpoint
```
Comment (karaimin) [inline scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/E2EGroupComplianceTest.java]: The 2 test bodies are very similar. Extract it

File: scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/E2EGroupComplianceTest.java [hunk 8]
```diff
@@ -913,17 +940,16 @@ public void testGetGroupsPagingStartIdEqEnd() {
...
-        () -> assertTrue(pagedGroupsResult.getTotalResults() > 0, "Verify 'totalResults' is greater than 0"),
+    assertAll("Verify List Response", () -> assertTrue(pagedGroupsResult.getTotalResults() > 0, "Verify 'totalResults' is greater than 0"),
         () -> assertTrue(Integer.parseInt(DEFAULT_COUNT) <= pagedGroupsResult.getItemsPerPage(), "Verify 'count' is equal or less to 'itemsPerPage'"),
         () -> assertEquals(startId, pagedGroupsResult.getStartId(), "Verify 'startId'"),
-        () -> assertEquals(PAGINATION_BY_ID_END_PARAM, pagedGroupsResult.getNextId(), "Verify 'nextId'")
-    );
+        () -> assertEquals(PAGINATION_BY_ID_END_PARAM, pagedGroupsResult.getNextId(), "Verify 'nextId'"));
   }
 
   @Test
```
Comment (karaimin) [inline scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/E2EGroupComplianceTest.java]: Enable for Empty Groups only

File: scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/E2EGroupComplianceTest.java [hunk 9]
```diff
@@ -1025,11 +1093,49 @@ public void testGetGroupsSeveralPagesUsingIdPaging() {
       PagedByIdentitySearchResult<Group> pagedGroups = groupFailSafeClient.getPagedById(startId, count);
 
       final String startIdCopy = startId;
-      assertAll("Verify List Response",
-          () -> assertEquals(startIdCopy, pagedGroups.getStartId(), "Verify 'startId'"),
+      assertAll("Verify List Response", () -> assertEquals(startIdCopy, pagedGroups.getStartId(), "Verify 'startId'"),
```
Comment (karaimin) [inline scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/E2EGroupComplianceTest.java]: surround with formatter

File: scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/E2EGroupComplianceTest.java [hunk 10]
```diff
@@ -913,17 +940,16 @@ public void testGetGroupsPagingStartIdEqEnd() {
...
-    SCIMResponse<PagedByIdentitySearchResult<Group>> pagedGroupsResponse = resourceAwareGroupRequest.readMultipleGroups(identityPageQuery().withStartId(startId));
+    SCIMResponse<PagedByIdentitySearchResult<Group>> pagedGroupsResponse = resourceAwareGroupRequest
+        .readMultipleGroups(identityPageQuery().withStartId(startId));
     assertTrue(pagedGroupsResponse.isSuccess(), "Verify GET response is success");
 
     PagedByIdentitySearchResult<Group> pagedGroupsResult = pagedGroupsResponse.get();
 
-    assertAll("Verify List Response",
-        () -> assertTrue(pagedGroupsResult.getTotalResults() > 0, "Verify 'totalResults' is greater than 0"),
+    assertAll("Verify List Response", () -> assertTrue(pagedGroupsResult.getTotalResults() > 0, "Verify 'totalResults' is greater than 0"),
```
Comment (karaimin) [inline scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/E2EGroupComplianceTest.java]: surround with formatter

File: scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/conditions/BackendSystemStateChecker.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import com.sap.scimono.scim.system.tests.SCIMComplianceTest;
+import com.sap.scimono.scim.system.tests.extensions.GroupClientScimResponseExtension;
+import com.sap.scimono.scim.system.tests.extensions.GroupFailSafeClient;
+import com.sap.scimono.scim.system.tests.extensions.UserClientScimResponseExtension;
+import com.sap.scimono.scim.system.tests.extensions.UserFailSafeClient;
+
+public class BackendSystemStateChecker extends SCIMComplianceTest {
```
Comment (karaimin) [inline scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/conditions/BackendSystemStateChecker.java]: package private

File: scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/conditions/BackendSystemStateChecker.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+  @RegisterExtension
+  GroupClientScimResponseExtension resourceAwareGroupRequest = GroupClientScimResponseExtension.forClearingAfterEachExecutions(groupRequest);
+
+  private final UserFailSafeClient userFailSafeClient = resourceAwareUserRequest.getFailSafeClient();
+  private final GroupFailSafeClient groupFailSafeClient = resourceAwareGroupRequest.getFailSafeClient();
+
+  public boolean isUsersEndpointEmpty() {
+    int alreadyCreatedUsersCount = userFailSafeClient.getAllWithIdPaging().size();
+    if (alreadyCreatedUsersCount == 0) {
```
Comment (karaimin) [inline scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/conditions/BackendSystemStateChecker.java]: single return statement

File: scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/conditions/BackendSystemStateChecker.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+  @RegisterExtension
+  UserClientScimResponseExtension resourceAwareUserRequest = UserClientScimResponseExtension.forClearingAfterEachExecutions(userRequest);
+
+  @RegisterExtension
+  GroupClientScimResponseExtension resourceAwareGroupRequest = GroupClientScimResponseExtension.forClearingAfterEachExecutions(groupRequest);
+
+  private final UserFailSafeClient userFailSafeClient = resourceAwareUserRequest.getFailSafeClient();
+  private final GroupFailSafeClient groupFailSafeClient = resourceAwareGroupRequest.getFailSafeClient();
+
+  public boolean isUsersEndpointEmpty() {
```
Comment (karaimin) [inline scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/conditions/BackendSystemStateChecker.java]: package-private

File: scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/conditions/BackendSystemStateChecker.java [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+    int alreadyCreatedUsersCount = userFailSafeClient.getAllWithIdPaging().size();
+    if (alreadyCreatedUsersCount == 0) {
+      return true;
+    }
+
+    return false;
+  }
+
+  public boolean isGroupsEndpointEmpty() {
```
Comment (karaimin) [inline scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/conditions/BackendSystemStateChecker.java]: same

File: scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/conditions/BackendSystemStateChecker.java [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+    int alreadyCreatedGroupsCount = groupFailSafeClient.getAllWithIdPaging().size();
+    if (alreadyCreatedGroupsCount == 0) {
```
Comment (karaimin) [inline scimono-compliance-tests/src/main/java/com/sap/scimono/scim/system/tests/conditions/BackendSystemStateChecker.java]: same

**Comment by karaimin** [comment]
Some things that I want to mention outside from the code comments

1. Tests which include testGet<suffix>..... and are only aplicable for starting with resource creation must be renamed to -> testCreateAndGet<suffix>. Such tests are for example: testGetUserWithGroupsAndMultiValueAttrs, testGetUsersFilteredGroups, etc...

2.Some of the tests should also has it duplicate variant for non-empty backend - such as

testGetUsersNegativeCountAndStartIndex
testGetUsersDefaultCountWithStartIndex
testGetGroupsTotalCountWithStartIndex
testGetGroupsWithStarIndexEqualTotalResults
testGetGroupsDefaultCountWithStartIndex

3. Many of the comments are similar including extract similar body for the tests for empty and non-empty backend and the multiple differences from the code formatter in places where there were semantically ordered blocks but the eclipse formatter inlined them. I reported only some of the places they occur. Can you re-check all the changed test files again and fix the other places not reported by me.
